gax/src/main/java/com/google/api/gax/batching/BatcherStats.java

```diff
@@ -0,0 +1,194 @@
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.batching;
+
+import static com.google.common.util.concurrent.MoreExecutors.directExecutor;
+
+import com.google.api.core.ApiFuture;
+import com.google.api.core.ApiFutureCallback;
+import com.google.api.core.ApiFutures;
+import com.google.api.core.SettableApiFuture;
+import com.google.api.gax.rpc.ApiException;
+import com.google.api.gax.rpc.StatusCode;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import javax.annotation.Nullable;
+
+/**
+ * This class keeps the statistics about failed operations(both at RPC and ElementT level) in {@link
+ * Batcher}. This provides the count of individual exception failure and count of each failed {@link
+ * StatusCode.Code} occurred in the batching process.
+ */
+class BatcherStats {
+
+  private final Map<Class, Integer> requestExceptionCounts = new ConcurrentHashMap<>();
+  private final Map<StatusCode.Code, Integer> requestStatusCounts = new ConcurrentHashMap<>();
+  private final AtomicInteger partialBatchFailures = new AtomicInteger(0);
+  private final Map<Class, Integer> entryExceptionCounts = new ConcurrentHashMap<>();
+  private final Map<StatusCode.Code, Integer> entryStatusCounts = new ConcurrentHashMap<>();
+
+  private final Object lock = new Object();
+
+  synchronized void recordBatchFailure(Throwable throwable) {
+    Class exceptionClass = throwable.getClass();
+
+    synchronized (lock) {
+      if (throwable instanceof ApiException) {
+        StatusCode.Code code = ((ApiException) throwable).getStatusCode().getCode();
+        exceptionClass = ApiException.class;
+
+        if (requestStatusCounts.containsKey(code)) {
+          Integer codeCount = requestStatusCounts.get(code);
+          requestStatusCounts.put(code, ++codeCount);
+
+        } else {
+          requestStatusCounts.put(code, 1);
+        }
+      }
+
+      if (requestExceptionCounts.containsKey(exceptionClass)) {
+        Integer exCount = requestExceptionCounts.get(exceptionClass);
+        requestExceptionCounts.put(exceptionClass, ++exCount);
+      } else {
+        requestExceptionCounts.put(exceptionClass, 1);
+      }
+    }
+  }
+
+  <T> void recordBatchElementsCompletion(List<SettableApiFuture<T>> batchElementResultFutures) {
+    final AtomicBoolean markBatchFailure = new AtomicBoolean();
+
+    for (ApiFuture<T> elementResult : batchElementResultFutures) {
+      ApiFutures.addCallback(
```

Comment:
I think try/catch would be easier to read:

```java
for (ApiFuture<T> elementFuture : batchElementResultFutures) {
  try {
    elementFuture.get();
  } catch (Throwable t) {
    ....
  }
}
```

Reply:
Also remove the atomic above

Reply:
If I understood correctly, you meant to block and check if any entry is being failed after `splitResponse()` is executed.   I believe going with this would create a lot of blocking time until each entry in the batch finishes applying the mutation. If possible I would like to avoid that.

I hope after the last commit, this method looks better, Can you please take a look and let me know your thoughts.

Reply:
All of the element results should be resolved at this point, so nothing should block. If an entry is not resolved at this point then it never will be. This would be due to a bug in the splitResponse. We can address this issue in a follow up PR that would iterate over all of the element results and sets an IllegalStateException for unresolved entries